server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,371 @@
+     *
+     * @param timeout The maximum time to wait for the bulk requests to complete
+     * @param unit    The time unit of the {@code timeout} argument
+     * @throws InterruptedException If the current thread is interrupted
+     */
+    public void awaitClose(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
+        synchronized (mutex) {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            this.cancellableFlushTask.cancel();
+
+            if (bulkRequest.numberOfActions() > 0) {
+                execute();
+            }
+            this.retry.awaitClose(timeout, unit);
+        }
+    }
+
+    /**
+     * Adds an {@link IndexRequest} to the list of actions to execute. Follows the same behavior of {@link IndexRequest}
+     * (for example, if no id is provided, one will be generated, or usage of the create flag).
+     */
+    public BulkProcessor2 add(IndexRequest request) {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds an {@link DeleteRequest} to the list of actions to execute.
+     */
+    public BulkProcessor2 add(DeleteRequest request) {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds either a delete or an index request.
+     */
+    public BulkProcessor2 add(DocWriteRequest<?> request) {
+        internalAdd(request);
+        return this;
+    }
+
+    protected void ensureOpen() {
+        if (closed) {
+            throw new IllegalStateException("bulk process already closed");
+        }
+    }
+
+    private void internalAdd(DocWriteRequest<?> request) {
+        // bulkRequest and instance swapping is not threadsafe, so execute the mutations under a mutex.
+        // once the bulk request is ready to be shipped swap the instance reference unlock and send the local reference to the handler.
+        Tuple<BulkRequest, Long> bulkRequestToExecute = null;
+        synchronized (mutex) {
+            ensureOpen();
+            bulkRequest.add(request);
+            bulkRequestToExecute = newBulkRequestIfNeeded();
+        }
+        // execute sending the local reference outside the lock to allow handler to control the concurrency via it's configuration.
+        if (bulkRequestToExecute != null) {
+            execute(bulkRequestToExecute.v1(), bulkRequestToExecute.v2());
+        }
+    }
+
+    private Scheduler.Cancellable startFlushTask(TimeValue flushInterval, ThreadPool threadPool) {
+        if (flushInterval == null) {
+            return new Scheduler.Cancellable() {
+                @Override
+                public boolean cancel() {
+                    return false;
+                }
+
+                @Override
+                public boolean isCancelled() {
+                    return true;
+                }
+            };
+        }
+        return threadPool.scheduleWithFixedDelay(() -> {
```

Comment:
Rather than polling for work to do every N seconds could we just schedule a flush each time we start adding items to a new bulk request? I reckon almost every time this executes will be a no-op.

Reply:
If we actually were using this to load a lot of data (not the ILMHistoryStore case), that could wind up scheduling a _lot_ of flushes. I guess we could keep track of whether one is scheduled, and only schedule something if one is not already scheduled.

Reply:
Yes something like that I think.